eksrollup/cli.py

```diff
@@ -11,32 +11,58 @@
+    cluster_health_wait = app_config['CLUSTER_HEALTH_WAIT']
     cluster_healthy = False
-    # check if asg has enough nodes first before checking instance health
-    if is_asg_scaled(asg_name, new_desired_asg_capacity):
-        # if asg is healthy start draining and terminating instances
-        if is_asg_healthy(asg_name):
-            # check if k8s nodes are all online
-            if k8s_nodes_count(desired_k8s_node_count):
-                # check k8s nodes are healthy
-                if k8s_nodes_ready():
-                    logger.info('Cluster validation passed. Proceeding with node draining and termination...')
-                    cluster_healthy = True
+    nodes_ready = False
+    retry_count = 0
+
+    while retry_count < cluster_health_retry:
+        retry_count += 1
+        if health_check_type == "asg":
+            logger.info(f'Waiting for {cluster_health_wait} seconds for ASG to scale before validating cluster health...')
+        else:
+            logger.info(f'Waiting for {cluster_health_wait} seconds before validating cluster health...')
+
+        time.sleep(cluster_health_wait)
+        desired_k8s_node_count = count_all_cluster_instances(cluster_name, predictive=predictive)
+
+        # check if asg has enough nodes first before checking instance health
+        if is_asg_scaled(asg_name, new_desired_asg_capacity):
+            # if asg is healthy start draining and terminating instances
+            if is_asg_healthy(asg_name):
+                # check if k8s nodes are all online and with the matches the desired amount of instances
+                if k8s_nodes_count(desired_k8s_node_count):
+                    # check k8s nodes are healthy
+                    if k8s_nodes_ready():
+                        nodes_ready = True
+                        cluster_healthy = True
+                        break
                 else:
-                    logger.info('Validation failed for cluster. Expected node count reached but nodes are not healthy.')
-            else:
-                nodes = get_k8s_nodes()
-                logger.info('Current k8s node count is {}'.format(len(nodes)))
-                logger.info('Validation failed for cluster. Current node count {} Expected node count {}.'.format(
-                    len(nodes),
-                    desired_k8s_node_count))
+                    node_count_mismatch = True
         else:
-            logger.info(
-                'Validation failed for asg {}.'
-                'Instances not healthy'.format(asg_name))
+            logger.info('Validation failed for asg {}. Not enough instances online.'.format(asg_name))
+
+    if cluster_healthy:
+        logger.info('Cluster validation passed. Proceeding with node draining and termination...')
+    elif not nodes_ready:
+        logger.info('Validation failed for cluster. Expected node count reached but nodes are not healthy.')
+    elif node_count_mismatch:
```

Comment:
Once the `node_count_mismatch` variable is set it to `True` is never cleared - and I think the order of the boolean checks are not quite right here and logging potentially log misleading state because it checks whether nodes are ready (and logs) **before** it checks `node_count_mismatch`, which is the opposite direction to how the flags are expected to flow above.

This means if, on say 1st try, the nodes count is **not** yet correct that it might log `Expected node count reached but nodes are not healthy` on the 2nd retry if the expected node count is still not actually yet reached.

To make this easier to reason about, I would suggest inverting `node_count_mismatch` to `node_count_expected` (so all the flags transition the same direction from `False` to `True` in normal operation) and make the node count check _before_ the `nodes_ready` check for the logging.

We could also possibly then remove `cluster_healthy` and instead just check `if not nodes_count_expected or not nodes_ready` to simplify the state management since the cluster health is essentially derived from these two variables.

What do you think?

Reply:
I see what you're saying with the order of the boolean checks and I fixed them as you mentioned. Good Catch!

The way I understand the next part is, during the below execution the first line 'k8s_nodes_count' checks if the desired node counts are reached and if they are not reached then its considered a mismatch from whats expected and whats currently there, so I would set the variable 'node_count_mismatch' to True, which I believe its named correctly since there is a mismatch.

       `        if k8s_nodes_count(desired_k8s_node_count): 
                    # check k8s nodes are healthy
                    if k8s_nodes_ready():
                        nodes_ready = True
                        cluster_healthy = True
                        break
                else:
                    node_count_mismatch = True`

I also added 'node_count_mismatch = False' to line 22 so that when the while loops runs again it sets back to False in case it does succeed this time. 

Since there was 2 variables set to True once it came down to the final if block below I decided that 2 were not needed as you mentioned but I ended up keeping just 'cluster_healthy' purely because of the name so that when someone else looks at the code I thought itd be easy for them to pinpoint where the health check is succeeding, but if you disagree with this then I can swap it out with 'nodes_ready = True'.

                    if k8s_nodes_ready():
                        cluster_healthy = True
                        break

Reply:
Thanks 👍 

It's a style thing, however I generally find it simpler to quickly understand algorithms that only switch boolean flags in one direction rather than toggling them back and forth; resetting at the start of a loop etc. And since the `cluster_healthy` uses `True` for target state (rather than being `cluster_unhealthy` personally I feel it makes it a bit harder to reason about than having two conditions that we both trying to get to `True` to consider the cluster healthy.

There are also still cases where it is misleading, if the ASG never scales for some reason and the instances never become ready (`is_asg_scaled(..) = False` or `is_asg_ready(..) = False`) it will exit the loop with `cluster_healthy = False; node_count_mismatch = False` but log `Validation failed for cluster. Expected node count reached but nodes are not healthy.` which is not known to be true, because it never actually got to the point of checking the nodes from the cluster perspective.

Additionally I don't think it's possible to get to the `else:` clause at the bottom now where it prints `Validation failed for asg BLAH. Instances not ready` because it cannot tell the difference between not healthy due to ASG health vs nodes not ready. It's also odd to then check `if not cluster_heathy` again before raising exception when the `elif` already has a clause checking for this.

I think this would be easier to get the logging correct by moving most of the logging back into the loop to avoid trying to infer the state afterwards based on a set of booleans that tell you which of the four nested `if`s you actually were able to check before existing the loop.

Reply:
I quickly tried my idea at https://github.com/crosbymichael1/eks-rolling-update/pull/1 (easier to express in code than text, possibly) to see what you think. Not tested, but will test if it's a direction we want to go.

Reply:
Just read through it and I like the format alot, I don't have to try and match the if/else blocks back and forth like before to see what happens if one if doesn't execute. It's all very simple and easy to read which I like.